uast/diff/diff.go

```diff
@@ -0,0 +1,287 @@
+				sum += ds.decideAction(src[i], dst[i]).cost()
+			}
+			return sum
+		}() == 0 {
+			bestDecision = min(bestDecision, sameDecision{basicDecision{cost}})
+			break
+		}
+
+		if len(src) != len(dst) {
+			cost = 2
+		}
+
+		if len(src) < len(dst) {
+			arr := make(nodes.Array, len(dst))
+			copy(arr, src)
+			src = arr
+		} else {
+			arr := make(nodes.Array, len(src))
+			copy(arr, dst)
+			dst = arr
+		}
+		n := len(src)
+		m := make([][]int, n)
+		for i := 0; i < n; i++ {
+			m[i] = make([]int, n)
+			for j := 0; j < n; j++ {
+				m[i][j] = ds.decideAction(src[i], dst[j]).cost()
+			}
+		}
+
+		res := lapjv.Lapjv(m)
+
+		for i1, i2 := range res.InRow {
+			if i1 != i2 {
+				cost = 2
+				break
+			}
+		}
+
+		cost += res.Cost
+
+		bestDecision = min(bestDecision, permuteDecision{basicDecision{cost}, res.InRow})
+
+	case nil:
+		cost = 0
+		bestDecision = min(bestDecision, sameDecision{basicDecision{cost}})
+
+	default:
+		panic(fmt.Errorf("unknown node type %T", src))
+	}
+
+	ds.decisions[label] = bestDecision
+	return ds.decisions[label]
+}
+
+func (ds *cacheStorage) push(change Change) {
+	ds.changes = append(ds.changes, change)
+}
+
+func (ds *cacheStorage) createRec(node nodes.Node) {
+	switch n := node.(type) {
+	case nodes.Object:
+		for _, v := range n {
+			ds.createRec(v)
+		}
+
+	case nodes.Array:
+		for _, v := range n {
+			ds.createRec(v)
+		}
+
+	default:
+		//values and nils are not saved separately
+		return
+	}
+	ds.push(Create{Node: node})
+}
+
+func (ds *cacheStorage) generateDifference(src, dst nodes.Node, parentID ID, parentKey Key) {
+	decision := ds.decideAction(src, dst)
+	switch d := decision.(type) {
+
+	//no action required if same
+	case sameDecision:
+
+	case replaceDecision:
+		//remove src (no action?) and create dst + attach it
+		if dst != nil {
+			ds.createRec(dst)
+			ds.push(Attach{Parent: parentID, Key: parentKey, Child: nodes.UniqueKey(dst)})
+		} else {
+			ds.push(Attach{Parent: parentID, Key: parentKey, Child: nil})
+		}
+
+	case matchDecision:
+		src, dst := src.(nodes.Object), dst.(nodes.Object)
+		keys := make(map[string]bool)
+		iterate := func(keyset nodes.Object) {
+			for key := range keyset {
+				if in := keys[key]; !in {
+					keys[key] = true
+					if _, ok := dst[key]; !ok {
+						ds.push(Deattach{Parent: nodes.UniqueKey(src), Key: String(key)})
+					} else {
+						ds.generateDifference(src[key], dst[key], nodes.UniqueKey(src), String(key))
+					}
+				}
+			}
+		}
+		iterate(src)
+		iterate(dst)
+
+	case permuteDecision:
+		src, dst := src.(nodes.Array), dst.(nodes.Array)
+		l := len(dst) - len(src)
+		//add possible nils to src
+		if l > 0 {
+			arr := make(nodes.Array, len(dst))
+			copy(arr, src)
+			src = arr
+		}
+		recreate := false
+		if l != 0 {
+			recreate = true
+		}
+		for i1, i2 := range d.permutation {
+			if i1 != i2 {
+				recreate = true
+				break
+			}
+		}
+		if recreate {
+			//recreate src with right perm
+			newsrc := make([]nodes.Node, len(dst))
```

Comment:
It's confusing that `newsrc` has the length of `dst` and iterates over `dst` index, but accesses `src` and `d.permutation` instead. If I'm not mistaken, here `src` has the same size, so `len(src)` can be used instead.

Reply:
No. `src` is/can be of different length than `dst`, that's why I'm creating a `newsrc` with length of `dst` and elements of `src` permuted by `d.permutation`.